compiler/env/newmemory/RawAllocator.hpp

```diff
@@ -0,0 +1,133 @@
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef TR_RAWALLOCATOR_INCL
+#define TR_RAWALLOCATOR_INCL
+
+#pragma once
+
+#include <stddef.h>  // for size_t
+#include <cstdlib>   // for free, malloc
+#include <new>       // for bad_alloc, nothrow, nothrow_t
+#include "env/TypedAllocator.hpp"
+
+
+namespace TR {
+
+typedef void *RawMemory;
+
+// RawAllocator is an abstract class used by the rest of the compiler to interact
+// with RawMemory allocators (e.g. concrete subclasses like MallocAllocator,
+// DebugRawAllocator, etc.). RawAllocator objects are expected to be thread-safe,
+// and to be *interchangeable*. Interchangeable means that objects *of a particular
+// subclass* are supposed to be like one another in that it should be perfectly valid
+// for one object to allocate a piece of RawMemory and for a different object to be used
+// to deallocate that RawMemory.
+//
+// It is not expected that objects from different subclasses of RawAllocator will be
+// interchangeable, since different subclasses may allocate/deallocate memory using
+// different facilities.
+//
+
+class RawAllocator
+   {
+public:
+   RawAllocator()
+      { }
+
+   // virtual "copy constructor" used primarily to simplify lifetime management for RawAllocator objects
+   virtual TR::RawAllocator & clone() = 0;
+
+   virtual RawMemory allocate(size_t size, const std::nothrow_t tag, void * hint = 0) const throw() =  0;
+   RawMemory allocate(size_t size, void * hint = 0) const
+      {
+      void * const alloc = allocate(size, std::nothrow, hint);
+      if (!alloc) throw std::bad_alloc();
+      return alloc;
+      }
+
+   virtual void deallocate(RawMemory p) const throw() = 0;
+
+   // default simplification: deallocate() does not need a size
+   virtual void deallocate(RawMemory p, const size_t size) const throw()
+      {
+      deallocate(p);
+      }
+
+   // Try to make the given RawMemory region inaccessible so future access will trap
+   // Not all subclasses will override this empty implementation
+   virtual void protect(RawMemory p, size_t size) const throw()
```

Comment:
The purpose of protect is so that memory allocation, deallocation, and protection can have consistent implementations in some particular RawAllocator. I suppose another option would be for *only* TR::DebugAllocator to have a protect function, but that may require upcasting somewhere else in the compiler from TR::RawAllocator. The intention was for the entire rest of the compiler to consistently use TR::RawAllocator as its interface.